### PR TITLE
Occurrence visualization

### PR DESCRIPTION
**Pull request by Blodir**
https://www.pivotaltracker.com/story/show/183185605
https://www.pivotaltracker.com/story/show/183186103
https://www.pivotaltracker.com/story/show/183185809
https://www.pivotaltracker.com/story/show/183220403

https://183185605.dev.laji.fi/observation/map

Notes:
* I'd like to hear opinion(s) about whether the legend should be moved up in hierarchy (so that it is not a child of laji-map, but laji-observation-map, or perhaps its own module?). I'm leaning on moving it to its own module and instanciating it in observation-map.

### CODE REVIEW

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts [hunk 1]
```diff
@@ -0,0 +1,299 @@
...
+const sliceYear = (newestRecord: string): number => (
+  parseInt((newestRecord).slice(0, 4), 10)
+);
+
+const getRecordAgeColor = (newestRecords: string | string[]): string => {
+  if (!newestRecords) { return '#ffffff'; }
+  if (!Array.isArray(newestRecords)) { return lajiMapObservationVisualization.recordAge.categories[getRecordAgeColorIdx(sliceYear(newestRecords))].color; }
+  const a = newestRecords.map(s => sliceYear(s));
+  const year = Math.max(...a);
+  return lajiMapObservationVisualization.recordAge.categories[getRecordAgeColorIdx(year)].color;
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:113]: Looks familiar. Similarly to the other color getters I commented about, this could be simplified a bit (treat input as array, use default value). Now that I've read this far, I wonder if these functions could be refactored to use some shared logic. There are maybe enough differences that it's not worth it...

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+      },
+      {
+        color: '#fdbf66',
+        //label: 'Uncertain'
+        translate: 'laji-map.legend.mode.recordQuality.4'
+      },
+      {
+        color: '#f26840',
+        //label: 'Erroneous'
+        translate: 'laji-map.legend.mode.recordQuality.5'
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:196]: This wouldn't need the commented labels if the translation keys would be constructed from enumerations instead of indices. But it's OK as it is.

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        label: `≥ ${currentYear - yearsAgoBreakpoints[2]}`
+      },
+      {
+        color: '#fdbf66',
+        label: `≥ ${currentYear - yearsAgoBreakpoints[3]}`
+      },
+      {
+        color: '#f26840',
+        label: `< ${currentYear - yearsAgoBreakpoints[3]}`
+      }
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:291]: Is the `yearsAgoBreakpoints[3]` twice on purpose here? This is why hard coded indices can produce bugs easily (first doing copy pasting and manually hard coding the changes is prone to typos, then after doing that a several times your eyes just see a repetitive pattern and don't notice that one of the indices is wrong. It's like trying to find Waldo). This could be more readable if for example there was a single data structure that would hold the year breakpoints and the colors.

**Review by olzraiti — CHANGES_REQUESTED** [review]
In general it's good stuff. I left some change request & other comments, main concern being the hard coded indices - the data structure could be improved.

**Comment by olzraiti** [comment]
> I'd like to hear opinion(s) about whether the legend should be moved up in hierarchy (so that it is not a child of laji-map, but laji-observation-map, or perhaps its own module?). I'm leaning on moving it to its own module and instanciating it in observation-map.

The legend is also used for line transect routes reservation state (https://beta.laji.fi/project/MHL.1/form), so it can't be a child of the observation map.  It's not that much code, so I don't mind it being inside the laji-map module. If you decide to move it to a separate module, that's also OK.

**Comment by olzraiti** [comment]
Maintenance wise, I think the most important thing to do would be to have the logic for map colors & legend to a single-source-of-truth data structure, so that when editing the visualization only a single place would have to be edited.

File: projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+export interface LajiMapVisualizationCategory { color: string; label?: string; translate?: string };
+export type LajiMapVisualization<T extends string> = Record<T,
+  {
+    label: string;
+    categories: LajiMapVisualizationCategory[];
+    getFeatureStyle?: (opt: GetFeatureStyleOptions) => PathOptions;
+    getClusterStyle?: (childCount: number, featureIdxs: number[], cluster: MarkerCluster) => PathOptions;
+    getClusterClassName?: (childCount: number, featureIdxs: number[], cluster: MarkerCluster) => string;
+  }
+>;
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts:13]: Fixed, also in relation to SSR, I noticed a nasty type error while building. I'm not sure if this is a bug in angular or what, but in order to keep onVisualizationModeChange typed I had to introduce this creature `(visualizationModeChange)="onVisualizationModeChange($any($event))"` in `observation-map.component.html`

File: projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { GetFeatureStyleOptions } from 'laji-map';
+import { MarkerCluster, PathOptions } from 'leaflet';
+
+export interface LajiMapVisualizationCategory { color: string; label?: string; translate?: string };
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts:4]: Idk how I feel about this. Practically there probably won't be any conflicts, but just the possibility that there could be one makes me feel a bit uneasy

File: projects/laji/src/app/shared-modules/laji-map/visualization/legend.component.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    <h5>{{ 'laji-map.legend.coordinateAccuracy' | translate }}<laji-info>{{ 'laji-map.legend.coordinateAccuracy.info' | translate }}</laji-info></h5>
+    <ul>
+      <li class="d-flex"><div class="legend-circle accuracy-1 mr-2"></div><div class="legend-label">1 - 10 m</div></li>
+      <li class="d-flex"><div class="legend-circle accuracy-2 mr-2"></div><div class="legend-label">10 - 100 m</div></li>
+      <li class="d-flex"><div class="legend-circle accuracy-3 mr-2"></div><div class="legend-label">100 - 1000 m</div></li>
+      <li class="d-flex"><div class="legend-circle accuracy-4 mr-2"></div><div class="legend-label">1000 - 10 000 m</div></li>
+      <li class="d-flex"><div class="legend-circle accuracy-5 mr-2"></div><div class="legend-label">10 000 - 100 000 m</div></li>
+    </ul>
+  </ng-container>
+</div>
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/laji-map/visualization/legend.component.html:31]: I can't think of any way to lift the coordinate accuracy from the legend without changing the UI... This is a big reason why I'm thinking about moving the legend to its own module.

File: projects/laji/src/app/shared-modules/laji-map/laji-map.component.ts [hunk 7]
```diff
@@ -30,35 +32,38 @@ import { environment } from 'projects/laji/src/environments/environment';
...
     </div>`,
   styleUrls: ['./laji-map.component.css'],
   providers: [],
   changeDetection: ChangeDetectionStrategy.OnPush
 })
 export class LajiMapComponent implements OnDestroy, OnChanges, AfterViewInit {
-
   @Input() data: any = [];
+  @Input() visualization: LajiMapVisualization<any> | undefined;
+  @Input() visualizationMode: string | undefined;
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/laji-map/laji-map.component.ts:50]: I originally had the component as generic, but ran into some issues with that (I forgot what exactly). I don't really see an issue with the generic type as at most it adds a little bit of boilerplate when using the type. I changed the `LajiMapVisualization<any>` to `LajiMapVisualization<string>` though.

File: projects/laji/src/app/shared-modules/laji-map/visualization/legend.component.html (hunk 6 above)
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/laji-map/visualization/legend.component.html:31]: We could provide a strategy name for rendering extra stuff, that would point to a template name. Similarly to how we handle datatable cell templates. So for example, `LajiMapVisualization` could have a prop `extraTemplate`, and `lajiMapObservationVisualization` could have something like:

```
const lajiMapObservationVisualization = {
...,
extraTemplate: 'coordinateAccuracy'
```

Then there'd be a template for `coordinateAccuracy` that is used. It's a bit ugly that the actual template implementation would have to be inside the `legend.component.html`, but Angular doesn't provide any alternatives.

My initial complaint was that the observation related code is inside the the legend template code, but at least this way it'd be more extendable and the API of the legend component wouldn't have anything related to the observation stuff (other than a template strategy name).

File: projects/laji/src/app/shared-modules/laji-map/laji-map.component.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/laji-map/laji-map.component.ts:50]: Ok, sounds reasonable

File: projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/laji-map/visualization/laji-map-visualization.ts:4]: I read the usage again, and I was confused because I didn't notice that there are two different `label` props (one for the visualization instance and then for the categories). So no problem here :)

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+      return 4;
+  }
+};
+
+const getRecordQualityColor = (qualities: string | string[]): string => {
+  if (!qualities) { return lajiMapObservationVisualization.recordQuality.categories[2].color; }
+  if (!Array.isArray(qualities)) { return lajiMapObservationVisualization.recordQuality.categories[getRecordQualityColorIdx(qualities)].color; }
+  const a = qualities.map(q => getRecordQualityColorIdx(q));
+  return lajiMapObservationVisualization.recordQuality.categories[Math.min(...a)].color;
+};
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:49]: When this was first implemented with clustering the array input was required. It's no longer used so I removed it. I'm not sure how much it improves the code, but I did this with the enum:
```
enum RecordQuality { ExpertVerified, CommunityVerified, Neutral, Uncertain, Erroneous }
const recordQualityMap = {
  EXPERT_VERIFIED: RecordQuality.ExpertVerified,
  COMMUNITY_VERIFIED: RecordQuality.CommunityVerified,
  UNCERTAIN: RecordQuality.Uncertain,
  ERRONEOUS: RecordQuality.Erroneous
};
const getRecordQualityColor = (recordQuality: string | undefined): string => {
  if (!recordQuality) { return lajiMapObservationVisualization.recordQuality.categories[RecordQuality.Neutral].color; }
  return lajiMapObservationVisualization.recordQuality.categories[recordQualityMap[recordQuality]].color;
};
```

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts (hunk 3 above)
Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:291]: `yearsAgoBreakpoints[3]` is twice on purpose.

Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts:291]: I made this a little clearer by moving the `currentYear - breakpoint` calculation to the `RECORD_AGE_BREAKPOINTS` array, but I'm not sure if there's anything else that can be done. The values can't be written directly into the string/enum even if we wanted to since it depends on current year (which is dynamic). Besides honestly I think the logic here is pretty clear?

Comment (olzraiti) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts]: I was wrong about the label, it's supposed to be 
`< 1982` like it is now. Sorry :D I'll mark this as resolved

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-visualization.ts]: Thanks, just adding the enums makes it more readable.

**Comment by Blodir** [comment]
Note that there are a few new changes since @esko-piirainen requested that instead of doing a count query we do a point query first, check if there's too many results, and do a box query if necessary. That is now implemented and testable.

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 9]
```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
-    if (outOfActiveBounds
-      || zoomedInAndNotShowingPoints
-      || zoomedOut
+    if (
+      !(this.showingIndividualPoints && activeBoundsContainViewport)
+      && (
+           // Zoom in
+           this.activeZoomThresholdLevel > curActiveZoomThresholdLevel
+           // Zoom out
+        || this.activeZoomThresholdLevel < curActiveZoomThresholdLevel
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:245]: It's not a biggie, but could be shortened to `this.activeZoomThresholdLevel !== curActiveZoomThresholdLevel`

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 10]
```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
+  resetTable() {
+    this.selectedObservationCoordinates = undefined;
+    this.cdr.detectChanges();
+    // Wait until next cycle so that height has time to adjust to the removal of the table
     setTimeout(() => {
-      this.reloading = false;
-    }, 200);
+      this.tableViewHeightOverride = undefined;
+      this.cdr.detectChanges();
+    });
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:289]: Did you try could this be done at `AfterViewChecked` so no need for timeout?

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 11]
```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
+        },
+        marker: {
+          icon: getPointIconStyle
+        },
+        total: data.total
+      }))
+    );
+  }
+
+  private onAggregateDataClick(args: any[]) {
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:363]: Typing could be improved

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 12]
```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
+          firstPage.features.push(...allPages[i].features);
+        }
+        // return the modified first page
+        return {
+          featureCollection: { type: firstPage.type, features: firstPage.features },
+          on: {
+            click: (...args) => this.onAggregateDataClick(args)
+          },
+          total: firstPage.total
+        };
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:420]: Collecting all the features with `allPages.reduce()` would be more clear (clear enough so that you wouldn't need the comments).

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 13]
```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
-          )).pipe(tap(() => {
-            this.showingIndividualPoints = false;
-          }));
+          return this.getBoxDataOptions$(query).pipe(tap(() => this.showingIndividualPoints = false));
         }
-      }));
+      }),
+      // retry on timeout
+      timeout(WarehouseApi.longTimeout * 3),
+      delay(100),
```
Comment (olzraiti) [inline projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:488]: What's the delay about? If it's necessary could you add some explanation into a comment for it.

**Review by olzraiti — CHANGES_REQUESTED** [review]
I had missed the changes to `projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts` since Github hid it because there were so many changes. It was pretty clear, good job on cleaning the view port change detection. I left some minor wishes.

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts (hunk 13 above)
Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:488]: I actually don't know, I didn't change that line because I didn't know why it was there (git shows it as new line for some reason)

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts [hunk 14]
```diff
@@ -107,84 +172,47 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
...
+        draw: false
+      },
+      zoom: 1.5,
+      draw: false,
+      tileLayerName: LajiMapTileLayerName.openStreetMap,
+      clickBeforeZoomAndPan: true
+    };
     if ((environment as any).observationMapOptions) {
-      this._mapOptions = {...this._mapOptions, ...(environment as any).observationMapOptions};
+      Object.assign(this.mapOptions, ...(environment as any).observationMapOptions);
```
[the remark this answers is not in the document]
Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:197]: Object reference is changed in the line above already, there's no reason to create another object (at small cost of performance) immediately after so mutation makes sense here.

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts (hunk 10 above)
Comment (Blodir) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:289]: Idk it might work, but I don't see why this would be an improvement? Calling detectChanges upon every view check is unnecessarily bad for performance, or alternatively there could be a flag to tell if it should be called on the next viewCheck, but that would complicate the code unnecessarily. An empty `setTimeout` isn't evil really, it just creates a callback in the js event loop. To be fair it's not at all obvious why it's required, but I think the alternatives are worse

Comment (olzraiti) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts]: It's not that bad as it is, we can go with this. I agree it would become more hard to follow with a flag. With the comment it's understandable.

In general I'm always worried about setTimeouts maintenance wise for many reasons. It's just a single tiny setTimeout which can seem innocent, but I don't like that it assumes that the table height is updated synchronously - if that changes this will break and it will be challenging to debug. It's difficult to find inline asynchronous operations in code. The `AfterViewChecked` should be guaranteened to run after the the actually is updated, even if it becomes async. Also, if there appear more setTimeouts, the order of execution can be really chaotic to follow.

File: projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts (hunk 14 above)
[the remark this answers is not in the document]
Comment (olzraiti) [reply projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts:196]: I think a reassignment is much more readable as the mutations done by any function are always pretty implicit and can be easily missed. The performance cost is really minimal...  But it's no big deal, I was just curious what was the reasoning.

**Comment by Blodir** [comment]
The legend is in its own module now and the coordinate accuracy is added via content projection inside the observation-map component.

**Review by olzraiti — APPROVED** [review]
:+1: great!
